### DISCUSSION TITLE
Symfony Mailer & SendGrid custom arguments

### DISCUSSION THREAD

**Discussion by abzolv** [symfony/symfony#42481] [Q&A]
SendGrid has a feature where you can send custom arguments along with an email. The custom arguments are added to the SendGrid webhook events that they send back to your endpoint.

It does not appear to be currently possible to add the custom arguments when using Symfony Mailer.

Am I missing something, or can it be added as an enhancement?

https://docs.sendgrid.com/for-developers/tracking-events/event#custom-arguments
https://github.com/sendgrid/sendgrid-php/blob/main/examples/helpers/mail/example.php#L46

**Comment by tarlepp** [comment]
You should check this repository - https://github.com/symfony/sendgrid-mailer - if there is some support for those or not.

And if you don't find anything there, then create an issue to main repository - https://github.com/symfony/symfony/issues - about that.

**Comment by javiereguiluz** [comment]
@abzolv I don't know if it's the same, but Symfony Mailer allows to add metadata and tags to emails. See https://symfony.com/doc/current/mailer.html#adding-tags-and-metadata-to-emails

**Reply by abzolv** [reply]
The issue is that the `SendgridApiTransport::getPayload` method of [Symfony SendGrid Bridge](https://github.com/symfony/sendgrid-mailer/blob/5.3/Transport/SendgridApiTransport.php) does not contain any code to handle `TagHeader` and `MetadataHeader` instances added to the email headers in Symfony Mailer. Whereas, `PostmarkApiTransport::getPayload` does contain code to convert those headers into Postmark tracking format.

It must just be an oversight because [SendGrid tracking](https://docs.sendgrid.com/for-developers/tracking-events/event#custom-arguments) has been a long-time feature of SendGrid.

**Reply by javiereguiluz** [reply]
You are right. I think it's a matter of adding to the SendGrid Bridge some code similar to this: https://github.com/symfony/mailgun-mailer/blob/afd868ec8509d9ef372d37aedc7a5ee24e7697c3/Transport/MailgunApiTransport.php#L125-L135

If you have some free time, it'd be great if you could submit a pull request with these changes for the SendGrid Bridge. These mailer bridges are maintained by the community because Symfony creators can't use and know all possible mailer services. That's why we need help from people like you who are experienced in each mailer. Thanks 🙏

**Comment by next-sentence** [comment]
custom_args worked fine for me with this line of code:

`$email->getHeaders()->add(new MetadataHeader('Color', 'blue'));`

and make sure you use api transport: `MAILER_DSN=sendgrid+api://KEY@default?region=REGION`
